pages/api/progress.py

```diff
@@ -0,0 +1,245 @@
+
+        Accessed via URL: /progress/<slug:sequence_slug>/<username>/<int:rank>/
+
+        .. code-block:: http
+
+            GET /api/progress/sequence1/user1/3 HTTP/1.1
+
+        responds
+
+        .. code-block:: json
+
+            {
+                "sequence_slug": "sequence1",
+                "rank": 3,
+                "viewing_duration": "00:00:30"
+            }
+        """
+        return super().retrieve(request, *args, **kwargs)
+
+    def destroy(self, request, *args, **kwargs):
+        """
+        Delete an EnumeratedProgress instance.
+
+        Accessed via URL: /progress/<slug:sequence_slug>/<username>/<int:rank>/
+
+        .. code-block:: http
+
+            DELETE /api/progress/sequence1/user1/3 HTTP/1.1
+
+        responds with a `204 No Content` status, indicating successful deletion.
+        """
+        return super().destroy(request, *args, **kwargs)
+
+    def update(self, request, *args, **kwargs):
+        """
+        Update an EnumeratedProgress instance.
+
+        Accessed via URL: /progress/<slug:sequence_slug>/<username>/<int:rank>/
+
+        .. code-block:: http
+
+            PATCH /api/progress/sequence1/user1/3 HTTP/1.1
+
+            {
+                "viewing_duration": "00:01:00"
+            }
+
+        responds
+
+        .. code-block:: json
+
+            {
+                "sequence_slug": "sequence1",
+                "rank": 3,
+                "viewing_duration": "00:01:00"
+            }
+        """
+        return super().update(request, *args, **kwargs)
+
+    @action(detail=True, methods=['post'])
+    def ping(self, request, *args, **kwargs):
+        """
+        Update the viewing duration of an EnumeratedProgress instance.
+
+        This endpoint is used to send 'pings' via POST requests to update the
+        viewing duration of the progress instance. It prevents too frequent updates
+        by checking the time since the last ping.
+
+        .. code-block:: http
+
+            POST /api/progress/sequence1/user1/3/ping HTTP/1.1
+
+        responds
+
+        .. code-block:: json
+
+            {
+                "sequence_slug": "sequence1",
+                "username": "user1",
+                "rank": 3,
+                "viewing_duration": "00:00:40",
+                "last_ping_time": "2023-11-13T04:56:04.075172Z"
+            }
+
+        """
+        return self._handle_ping_viewing(request, *args, **kwargs)
+
+    def _handle_ping_viewing(self, request, *args, **kwargs):
+        """
+        Handle the logic for the ping action.
+        time_increment sets the minimum amount of pings received
+        """
+        instance = self.get_object()
+        now = timezone.now()
+        time_increment = timedelta(seconds=15)
+
+        if instance.last_ping_time:
+            time_since_last_ping = now - instance.last_ping_time
+            if time_since_last_ping >= time_increment:
+                instance.viewing_duration += time_increment
```

Comment:
interesting. we increment the viewing_duration by `time_increment` and not `time_since_last_ping`.